rule_builder.py

```diff
@@ -0,0 +1,1674 @@
+            for rule_id in rule_ids:
+                player_results.pop(rule_id, None)
+
+        return changed
+
+    @override
+    def remove(self, state: "CollectionState", item: "Item") -> bool:
+        changed = super().remove(state, item)
+        if not changed:
+            return changed
+
+        if getattr(self, "rule_item_dependencies", None):
+            player_results: dict[int, bool] = state.rule_cache[self.player]
+            mapped_name = self.item_mapping.get(item.name, "")
+            rule_ids = self.rule_item_dependencies[item.name] | self.rule_item_dependencies[mapped_name]
+            for rule_id in rule_ids:
+                player_results.pop(rule_id, None)
+
+        # clear all region dependent caches as none can be trusted
+        if getattr(self, "rule_region_dependencies", None):
+            for rule_ids in self.rule_region_dependencies.values():
+                for rule_id in rule_ids:
+                    state.rule_cache[self.player].pop(rule_id, None)
+
+        # clear all location dependent caches as they may have lost region access
+        if getattr(self, "rule_location_dependencies", None):
+            for rule_ids in self.rule_location_dependencies.values():
+                for rule_id in rule_ids:
+                    state.rule_cache[self.player].pop(rule_id, None)
+
+        # clear all entrance dependent caches as they may have lost region access
+        if getattr(self, "rule_entrance_dependencies", None):
+            for rule_ids in self.rule_entrance_dependencies.values():
+                for rule_id in rule_ids:
+                    state.rule_cache[self.player].pop(rule_id, None)
+
+        return changed
+
+    @override
+    def reached_region(self, state: "CollectionState", region: "Region") -> None:
+        super().reached_region(state, region)
+        if getattr(self, "rule_region_dependencies", None):
+            player_results: dict[int, bool] = state.rule_cache[self.player]
+            for rule_id in self.rule_region_dependencies[region.name]:
+                player_results.pop(rule_id, None)
+
+
+Operator = Literal["eq", "ne", "gt", "lt", "ge", "le", "contains"]
+
+OPERATORS = {
+    "eq": operator.eq,
+    "ne": operator.ne,
+    "gt": operator.gt,
+    "lt": operator.lt,
+    "ge": operator.ge,
+    "le": operator.le,
+    "contains": operator.contains,
+}
+
+T = TypeVar("T")
+TWorld = TypeVar("TWorld", bound=RuleWorldMixin, contravariant=True, default=RuleWorldMixin)  # noqa: PLC0105
+
+
+@dataclasses.dataclass(frozen=True)
+class OptionFilter(Generic[T]):
+    option: "type[Option[T]]"
+    value: T
+    operator: Operator = "eq"
+
+    def to_json(self) -> dict[str, Any]:
+        return {
+            "option": f"{self.option.__module__}.{self.option.__name__}",
+            "value": self.value,
+            "operator": self.operator,
+        }
+
+    @classmethod
+    def from_json(cls, data: dict[str, Any]) -> Self:
```

Comment:
I don't like the names `to_json` and `from_json` if they don't actually do the encoding and decoding.

The name `as_dict` is already used for this type of thing both in AP (`CommonOptions`) and outside of AP in standard Python.

I'm not sure if the name for the other direction is established the same amount, but I think it could be `from_dict`.

Reply:
Good point, I've renamed everything to `to_dict` and made the docs more clear about the world dev being responsible for the serialization